{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import pysam\n",
    "import numpy.random as r\n",
    "import re\n",
    "import ast\n",
    "from cyvcf2 import VCF\n",
    "from tqdm import tqdm\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Creates ITERATIONS number of fasta sequences from vcf2fasta sequences\n",
    "that have RECOMBINATION_CHANCE recombination rate for power analysis\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "args = {}\n",
    "# properties of vcf2fasta generated sequences sequences\n",
    "VCF2FASTA_START = 1\n",
    "VCF2FASTA_END = 800000\n",
    "VCF2FASTA_FILEPATH = 'output/vcf2fasta.fasta'\n",
    "\n",
    "# recombination event probabilitiy per base\n",
    "RECOMBINATION_CHANCE = 0.00004\n",
    "\n",
    "# number of recombined sequences to generate\n",
    "ITERATIONS = 5\n",
    "\n",
    "# tsv output filepath containing phase change locations for each sequence\n",
    "PHASE_CHANGE_TSV = 'phase_changes.tsv'\n",
    "# fasta output filepath containing all recombined sequences\n",
    "SEQUENCES_OUTPUT = 'generated_sequences.fasta'\n",
    "\n",
    "\n",
    "\n",
    "# parse results from vcf2fasta\n",
    "fasta_file = list(SeqIO.parse(VCF2FASTA_FILEPATH, 'fasta'))\n",
    "\n",
    "# tsv for phase changes\n",
    "phase_change_tsv = open(PHASE_CHANGE_TSV, 'w')\n",
    "\n",
    "# regex to identify correct fasta reference_name\n",
    "chrom = re.search(r'chromosome_[0-9]{1,2}', str(fasta_file[0].id)).group()\n",
    "\n",
    "fastas = []\n",
    "\n",
    "iteration_counter = 1\n",
    "\n",
    "for i in range(ITERATIONS):\n",
    "\n",
    "    events = r.poisson((VCF2FASTA_END - VCF2FASTA_START) * RECOMBINATION_CHANCE)\n",
    "    phase_changes = sorted(list(r.choice((VCF2FASTA_END - VCF2FASTA_START), size=events)))\n",
    "    \n",
    "    phase_change_tsv.write('sequence' + str(iteration_counter) + '\\t' + str(phase_changes) + '\\n')\n",
    "\n",
    "    haplotype0 = str(fasta_file[0].seq)\n",
    "    haplotype1 = str(fasta_file[1].seq)\n",
    "\n",
    "    current = 0\n",
    "    idx = VCF2FASTA_START\n",
    "    recomb_seq = []\n",
    "\n",
    "    for base in zip(haplotype0, haplotype1):\n",
    "\n",
    "        # switch haplotype if idx in list phase_changes    \n",
    "        current = (current + 1) % 2 if idx in phase_changes else current\n",
    "        recomb_seq.append(base[current])\n",
    "\n",
    "        if idx >= VCF2FASTA_END:\n",
    "            break\n",
    "        else:\n",
    "            idx += 1\n",
    "\n",
    "    recomb_record = SeqRecord(\n",
    "        Seq(''.join(recomb_seq)),\n",
    "        id=chrom + '_' + str(iteration_counter),\n",
    "        description=chrom\n",
    "    )\n",
    "\n",
    "    fastas.append(recomb_record)\n",
    "    iteration_counter += 1\n",
    "\n",
    "SeqIO.write(fastas, SEQUENCES_OUTPUT, 'fasta')\n",
    "phase_change_tsv.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This cell uses the output from above (SEQUENCES_OUTPUT) and generates read pairs using ART,\n",
    "then it uses readcomb.filter (not the one in pypi package, but the filter.py file in this folder) to\n",
    "generate statistics comparing number of filtered and unfiltered reads in windows of 1000\n",
    "\"\"\"\n",
    "\n",
    "import subprocess\n",
    "from cyvcf2 import VCF\n",
    "from tqdm import tqdm\n",
    "\n",
    "# import classification new one I haven't uploaded to pypi\n",
    "# not good habit so don't do this if you're copying this script\n",
    "from classification import *\n",
    "\n",
    "\n",
    "VCF_FILEPATH = 'data/filtered_full.vcf.gz'\n",
    "vcf = VCF(VCF_FILEPATH)\n",
    "\n",
    "window_counts_tsv = open('window_counts.tsv', 'w')\n",
    "\n",
    "COVERAGE = 0.25\n",
    "INSERT = 100\n",
    "UNFILTERED_OUTPUT = 'unfiltered'\n",
    "FILTERED_OUTPUT = 'filtered'\n",
    "\n",
    "iteration = 1\n",
    "\n",
    "for i in range(5):\n",
    "    \n",
    "    # run art_illumina on 400 sequences with coverage 400\n",
    "    print('Running art_illumina iteration: ' + str(iteration))\n",
    "    \n",
    "    subprocess.run(['art_illumina',\n",
    "                    '-f', str(COVERAGE),\n",
    "                    '-l', '250',\n",
    "                    '-ss', 'MSv1',\n",
    "                    '-i', SEQUENCES_OUTPUT,\n",
    "                    '-m', str(500 + INSERT),\n",
    "                    '-s', '1',\n",
    "                    '-o', UNFILTERED_OUTPUT + str(iteration),\n",
    "                    '-M', '-p', '-sam', '-na', '-q'])\n",
    "\n",
    "    # run readcomb on 400 sequence output generated by art_illumina\n",
    "    print('Running readcomb iteration: ' + str(iteration))\n",
    "\n",
    "    subprocess.run(['readcomb-filter',\n",
    "                    '-b', UNFILTERED_OUTPUT + str(iteration) + '.sam',\n",
    "                    '-v', VCF_FILEPATH,\n",
    "                    '-p', '4',\n",
    "                    '-o', FILTERED_OUTPUT + str(iteration)])\n",
    "    \n",
    "    \n",
    "\n",
    "    # read in sam files and \n",
    "    \n",
    "    unfiltered_pairs = pairs_creation(UNFILTERED_OUTPUT + str(iteration) + '.sam', VCF_FILEPATH)\n",
    "    filtered_pairs = pairs_creation(FILTERED_OUTPUT + str(iteration) + '.sam', VCF_FILEPATH)\n",
    "    \n",
    "    # call get_midpoint on all of them, this will take quite a long time so use tqdm\n",
    "    \n",
    "    #[[filtered, unfiltered], ...]\n",
    "    window_counts = [[0,0] for i in range(200)]\n",
    "    \n",
    "    for pair in tqdm(filtered_pairs):\n",
    "        pair.classify(vcf=vcf)\n",
    "        midpoint = pair.get_midpoint()\n",
    "        window = int(midpoint // 1000)\n",
    "        window_counts[window][0] += 1    \n",
    "    \n",
    "    for pair in tqdm(unfiltered_pairs):\n",
    "        start = pair.rec_1.reference_start\n",
    "        end = pair.rec_2.reference_start + pair.rec_2.query_alignment_length\n",
    "        midpoint = int((start + end) / 2)\n",
    "        \n",
    "        window = int(midpoint // 1000)\n",
    "        window_counts[window][1] += 1\n",
    "    \n",
    "    for i in range(len(window_counts)):\n",
    "        window_counts_tsv.write('\\t'.join(map(str, [\n",
    "            iteration,\n",
    "            i * 1000,\n",
    "            (i + 1) * 1000,\n",
    "            window_counts[i][0],\n",
    "            window_counts[i][1]\n",
    "        ])) + '\\n')\n",
    "        \n",
    "    iteration += 1\n",
    "\n",
    "window_counts_tsv.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
