{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is for developing my new approach to VCF filtering\n",
    "Because the attributes of VCFs created with different callers vary\n",
    "I've made a new filtering strategy that works with any set of arbitrary filters\n",
    "the state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#PopGenVCF\n",
    "\n",
    "import vcf, sys, re\n",
    "from collections import Counter\n",
    "from ness_vcf import SFS\n",
    "\n",
    "\n",
    "def test_site(pyVCF_RecordObject, recordAttribute_filters, recordInfo_filters, lenient=True, verbose = False):\n",
    "    \"\"\"\n",
    "    this will filter based on whole site categories\n",
    "    (1)Attributes of the record that can be accessed like this \"record.ATTRIBUTE\"\n",
    "        e.g. QUAL, POS, FILTER, REF, ALT,  num_het, num_called, is_indel, is_sv etc\n",
    "    (2)INFO field filters accessed like this record.INFO[field[\n",
    "        e.g. MQ, DP, AF, AN etc\n",
    "\n",
    "    THE challenge with whole SITE filters is that they won't necessarily all be numbers:\n",
    "        e.g. we may want a site with QUAL > 30 and MQ > 20 BUT with ReF\n",
    "    We need the filter dict to include the name of the thing plus the test\n",
    "    then use eval to do the test:\n",
    "        eg. f = 'QUAL': '> 30.0'\n",
    "            if not eval('pyVCF_RecordObject.%s %s' %(f, SiteProperty_filter_dict[f]))\n",
    "            #which is thq equivalent of eval(record.QUAL > 30)\n",
    "    \"\"\" \n",
    "    for f in recordAttribute_filters:\n",
    "        try:\n",
    "            if not eval('pyVCF_RecordObject.%s %s' %(f, recordAttribute_filters[f])): #this means that along the way if any of the filters fail its over for this call\n",
    "                return False\n",
    "        except AttributeError:\n",
    "            if verbose: sys.stderr.write(\"%s does not exist in record at site %s:%i \\n\" %(f, pyVCF_RecordObject.CHROM, pyVCF_RecordObject.POS))\n",
    "            if lenient: continue\n",
    "            else: return False\n",
    "    for f in recordInfo_filters:\n",
    "        try:\n",
    "            if not eval(\"pyVCF_RecordObject.INFO['%s'] %s\" %(f, recordInfo_filters[f])): #this means that along the way if any of the filters fail its over for this call\n",
    "                return False\n",
    "        except KeyError:\n",
    "            if verbose: sys.stderr.write(\"%s does not exist in the INFO field of site %s:%i \\n\" %(f, pyVCF_RecordObject.CHROM, pyVCF_RecordObject.POS))\n",
    "            if lenient: continue\n",
    "            else: return False\n",
    "    #if a site got here without returning false its a keeper\n",
    "    return True\n",
    "\n",
    "            \n",
    "def test_genotype(pyVCF_CallObject, callFormat_filters, callAttribute_filters, lenient=True, verbose =False):\n",
    "    \"\"\"\n",
    "    we want this function to take a PyVCF call object and test whether it passes all the defined filters\n",
    "    if that field doesn't exist in this call object either return fail or go onto the other filters\n",
    "    filters = {'min_GQ': min_GQ, 'min_DP': min_DP, 'min_RGQ': min_RGQ, 'min_purity': min_purity, 'called': called, ' heterozygote': heterozygote}\n",
    "    # ???how do we translate the filters you want to a series of tests ???\n",
    "        so things in the genotype field will be stored in a dictionary\n",
    "            eg call['GT'] = 0, call['RGQ'] = 99 #need to be sure these are all minimum values\n",
    "            these things can be accessed using the filter name to get the value\n",
    "        other filters that are tests can be accessed from pyVCFs built in convenience functions\n",
    "            eg. call.is_snp, call.called - NEED TO BE CAREFUL to ensure these are all True/FALSE\n",
    "            to access these we can use the filter name eg 'called'  or 'is_het' \n",
    "            along with the 'getattr(c, filter)' to return the value\n",
    "    perhaps the best way to set this up is to pass two dicts:\n",
    "        (1) GT call filters ie anything that might show up in the record field\n",
    "        (2) pyVCF attributes \n",
    "    \"\"\"\n",
    "    #####################\n",
    "    if not pyVCF_CallObject.called:\n",
    "        return False\n",
    "    for f in callFormat_filters:\n",
    "        try:\n",
    "            if not eval(\"pyVCF_CallObject['%s'] %s\" %(f, callFormat_filters[f])): #this means that along the way if any of the filters fail its over for this call\n",
    "               return False\n",
    "        except AttributeError: #this catches cases where the field doesn't exist in that call - \n",
    "            if verbose: sys.stderr.write(\"%s does not exist in genotype %s at site %s:%i \\n\" %(f, pyVCF_CallObject.sample,c.site.CHROM, c.site.POS))\n",
    "            if lenient : continue\n",
    "            else: return False\n",
    "    for f in callAttribute_filters:\n",
    "        try:\n",
    "            if not eval('pyVCF_CallObject.%s %s' %(f, callAttribute_filters[f])): #this means that along the way if any of the filters fail its over for this call\n",
    "                return False\n",
    "        except AttributeError: #this catches cases where the field doesn't exist in that call - \n",
    "            if verbose: sys.stderr.write(\"%s is not an attribute of genotype %s at site %s:%i \\n\" %(f, pyVCF_CallObject.sample,c.site.CHROM, c.site.POS))\n",
    "            if lenient : continue\n",
    "            else: return False  \n",
    "    # if you've reached here without returning False the site must have passed\n",
    "    return True\n",
    "\n",
    "def genotypes_to_MAF(record, recordAttribute_filters=None, recordINFO_filters=None, callFormat_filters=None, callAttribute_filters=None, samples='all', site_lenient=True, call_lenient=True, ploidy=2, vcf_ploidy=2, minHQ_GTs =2 ):\n",
    "\t##########################################################################################\n",
    "\t\"\"\" takes record and returns minor allele freq and num_called which exceed quality threshold \"\"\"\n",
    "\t##########################################################################################\n",
    "\tif record.is_monomorphic and test_site(record,recordAttribute_filters, recordINFO_filters, lenient = site_lenient):\n",
    "\t\tif samples == 'all' or samples == ['all']:\n",
    "\t\t\tGTs = [i['GT'] for i in record.samples if test_genotype(i,callFormat_filters, callAttribute_filters, lenient=call_lenient)]\n",
    "\t\telse:\n",
    "\t\t\tGTs = [record.genotype(i)['GT'] for i in samples if test_genotype(i,callFormat_filters, callAttribute_filters, lenient=call_lenient)]\n",
    "\t\tif len(GTs) > minHQ_GTs:\n",
    "\t\t\treturn (0.0, len(GTs)*ploidy)\n",
    "\t\telse: return None\n",
    "\telif test_site(record,recordAttribute_filters, recordINFO_filters, lenient = site_lenient) and not record.is_monomorphic:\n",
    "\t\tif samples == 'all' or samples == ['all']:\n",
    "\t\t\tGTs = [i['GT'] for i in record.samples if test_genotype(i,callFormat_filters, callAttribute_filters, lenient=call_lenient)]\n",
    "\t\telse:\n",
    "\t\t\tGTs = [record.genotype(i)['GT'] for i in samples if test_genotype(i,callFormat_filters, callAttribute_filters, lenient=call_lenient)]\n",
    "\t\t# now make some calculations\n",
    "\t\ttry:\n",
    "\t\t\tif len(GTs) < minHQ_GTs: return None\n",
    "\t\t\telif ploidy == 2: \n",
    "\t\t\t\tMAF = Counter(re.findall('[0-9]', str(GTs))).most_common(2)[-1][-1]/(2.0*len(GTs))\n",
    "\t\t\t\treturn (MAF, 2*len(GTs))\n",
    "\t\t\telif ploidy == 1 and vcf_ploidy==2:\n",
    "\t\t\t\tMAF = Counter(re.findall('[0-9]', str(GTs))).most_common(2)[-1][-1]/(2.0*len(GTs))\n",
    "\t\t\t\treturn (MAF, len(GTs))\n",
    "\t\t\telif ploidy==1 and vcf_ploidy==1:\n",
    "\t\t\t\tMAF = Counter(re.findall('[0-9]', str(GTs))).most_common(2)[-1][-1]/(1.0*len(GTs))\n",
    "\t\t\t\treturn (MAF, len(GTs))\n",
    "\t\t\telse: return None\n",
    "\t\texcept ZeroDivisionError:\n",
    "\t\t\treturn None\n",
    "\telse:\n",
    "\t\treturn None\n",
    "\t# #if the site is invariant we only use depth filters\n",
    "\t# elif 'MQ' in record.INFO and record.num_called >= min_called and record.INFO['MQ']>=min_MQ and not indel(record):\n",
    "\t# \tif samples == 'all' or samples == ['all']: callable = len([i for i in record.samples if (i and i['DP']>min_DP)])\n",
    "\t# \telse: callable = len([record.genotype(i)['DP'] for i in samples if (record.genotype(i)['DP']>min_DP)])\n",
    "\t# \tif callable >= min_called:\n",
    "\t# \t\treturn (0.0, callable*ploidy)\n",
    "\t# elif record.num_called >= min_called and not indel(record):\n",
    "\t# \tif samples == 'all' or samples == ['all']: callable = len([i for i in record.samples if (i and i['DP']>min_DP)])\n",
    "\t# \telse: callable = len([record.genotype(i)['DP'] for i in samples if (record.genotype(i)['DP']>min_DP)])\n",
    "\t# \tif callable >= min_called:\n",
    "\t# \t\treturn (0.0, callable*ploidy)\n",
    "\t# else: \n",
    "\t# \treturn None\n",
    "\n",
    "def SFS_from_genotypes(vcf_reader, region, recordAttribute_filters=None, recordINFO_filters=None, callFormat_filters=None, callAttribute_filters=None, samples='all', site_lenient=True, call_lenient=True, ploidy=2, vcf_ploidy=2,minHQ_GTs=2):\n",
    "\t\"\"\"this goes to each site in a VCF and calculates the alternate allele frequency then summarises those calculations in SFS objects\n",
    "\tIt returns a dictionary of SFSs - where the key is the number of alleles and the value is the SFS of sites with that many alleles\"\"\"\n",
    "\tchromosome = str(region.split(\":\")[0])\n",
    "\tstart_coord = int(region.split(\":\")[1].split(\"-\")[0])\n",
    "\tend_coord = int(region.split(\":\")[1].split(\"-\")[1])\n",
    "\tAFs =[genotypes_to_MAF(record, recordAttribute_filters=recordAttribute_filters, recordINFO_filters=recordINFO_filters, callFormat_filters=callFormat_filters, callAttribute_filters=callAttribute_filters, samples=samples, site_lenient=site_lenient, call_lenient=call_lenient, ploidy=ploidy, vcf_ploidy=vcf_ploidy,minHQ_GTs=minHQ_GTs) \\\n",
    "\t\t\tfor record in vcf_reader.fetch(chromosome, start_coord, end_coord) \\\n",
    "\t\t\tif genotypes_to_MAF(record, recordAttribute_filters=recordAttribute_filters, recordINFO_filters=recordINFO_filters, callFormat_filters=callFormat_filters, callAttribute_filters=callAttribute_filters, samples=samples, site_lenient=site_lenient, call_lenient=call_lenient, ploidy=ploidy, vcf_ploidy=vcf_ploidy,minHQ_GTs=minHQ_GTs) !=None]\n",
    "\tSFSs={}\n",
    "\tif samples == ['all'] or samples == 'all':\n",
    "\t\tfor i in range(minHQ_GTs,ploidy*len(vcf_reader.samples)+1):\n",
    "\t\t\tSFSs[i] = SFS([0]*(i+1))\n",
    "\telse:\n",
    "\t\tfor i in range(minHQ_GTs, ploidy*len(samples)+1):\n",
    "\t\t\tSFSs[i] = SFS([0]*(i+1))\n",
    "\tfor a in AFs:\n",
    "\t\tnum_alleles = a[1]\n",
    "\t\tfreq = a[0]\n",
    "\t\tSFSs[num_alleles].add(freq, num_alleles)\n",
    "\treturn SFSs #dictionary of SFSs for multiple allele depths\n",
    "\n",
    "\n",
    "##### test a few things\n",
    "# region = 'chromosome_1:1000000-1010000'\n",
    "# vcf_file = '/Volumes/Ness_4TB/sequencing/chlamydomonas/quebec/all/haplotypeCaller/chromosome_1_4_development/chromosome_1.1000000-1100000.bpRES.vcf.gz'\n",
    "# vr = vcf.Reader(filename=vcf_file)\n",
    "\n",
    "# recordAttribute_filters={\"num_called\": \">8\", \"num_hets\": \"==0\"}\n",
    "# recordINFO_filters={}\n",
    "# callFormat_filters={\"GQ\": \">30\", \"RGQ\": \">30\", \"DP\":\">5\"}\n",
    "# callAttribute_filters={}\n",
    "# SFSs = SFS_from_genotypes(vr, region, recordAttribute_filters=recordAttribute_filters, recordINFO_filters=recordINFO_filters, callFormat_filters=callFormat_filters, callAttribute_filters=callAttribute_filters,  ploidy=1, vcf_ploidy=1)\n",
    "\n",
    "# SFSs[10].summarise()\n",
    "\n",
    "\n",
    "def vcf_stat_histo(vr, region, stat, bin_width, samples = 'all'):\n",
    "    chromosome = str(region.split(\":\")[0])\n",
    "    start_coord = int(region.split(\":\")[1].split(\"-\")[0])\n",
    "    end_coord = int(region.split(\":\")[1].split(\"-\")[1])\n",
    "    histo = OrderedDict()\n",
    "    for r in vr.fetch(chromosome, start_coord, end_coord):\n",
    "        r.stat\n",
    "    AFs =[genotypes_to_MAF(record, recordAttribute_filters=recordAttribute_filters, recordINFO_filters=recordINFO_filters, callFormat_filters=callFormat_filters, callAttribute_filters=callAttribute_filters, samples=samples, site_lenient=site_lenient, call_lenient=call_lenient, ploidy=ploidy, vcf_ploidy=vcf_ploidy,minHQ_GTs=minHQ_GTs) \\\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "hapvcf = vcf.Reader(filename='/Volumes/Ness_4TB/sequencing/chlamydomonas/quebec/all/haplotypeCaller/chromosome_1_4_development/chromosome_1.1000000-1100000.vcf.gz')\n",
    "#freevcf =vcf.Reader(filename='/Volumes/Ness_4TB/sequencing/chlamydomonas/quebec/all/freebayes/chromosome_1.dev/quebec_all.chromosome_1_1000000_1100000.vcf.gz') \n",
    "\n",
    "### This vcf is made with max-complex-gap ==0  which means that there shouldn't be any TYPE, snp,snp sites - ie more or less one site per line (except indels)\n",
    "freevcf =vcf.Reader(filename='/Volumes/Ness_4TB/sequencing/chlamydomonas/quebec/all/freebayes/chromosome_1.dev/quebec_all.chromosome_1_1000000_1100000.freebayes.bpRES.vcf.gz') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "recordAttribute_filters={\"num_called\": \">8\", \"num_hets\": \"==0\"}\n",
    "recordINFO_filters={}\n",
    "callFormat_filters={\"DP\":\">5\", \"GQ\" : \">0\", \"RGQ\" : \">00\"}\n",
    "callAttribute_filters={}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0, 0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0, 0, 0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "10\tFalse\t2\t1\t1\t0.266666666667\t0.17674288119\t1.3026779048\t[0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1]\n",
      "11\tFalse\t1\t0\t1\t0.509090909091\t0.341417152147\t1.18559726194\t[0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "15\tFalse\t8\t4\t4\t0.164285714286\t0.153772233094\t0.219180201508\t[1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 3]\n",
      "16\tFalse\t9\t6\t3\t0.125\t0.100455192819\t0.708323469178\t[4, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2]\n",
      "17\tFalse\t66\t62\t4\t0.0204991087344\t0.0179269207107\t0.439088315479\t[54, 0, 2, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 8]\n",
      "18\tFalse\t2657\t2505\t152\t0.00790119083639\t0.0166322149071\t-2.22012569387\t[2495, 135, 4, 2, 6, 1, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10]\n",
      "19\tFalse\t121\t98\t23\t0.0543231356628\t0.0543853409896\t-0.00445199336254\t[91, 3, 10, 1, 2, 1, 3, 2, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 7]\n",
      "20\tFalse\t6251\t5704\t547\t0.019540452475\t0.0246652819814\t-0.868736138935\t[5632, 209, 122, 75, 55, 16, 32, 16, 10, 4, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 72]\n"
     ]
    }
   ],
   "source": [
    "freeSFSs = SFS_from_genotypes(freevcf, region, recordAttribute_filters=recordAttribute_filters, recordINFO_filters=recordINFO_filters, callFormat_filters=callFormat_filters, callAttribute_filters=callAttribute_filters,  ploidy=1, vcf_ploidy=1)\n",
    "for s in freeSFSs:\n",
    "    freeSFSs[s].summarise(header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\tFalse\t0\t0\t0\tNone\tNone\tNone\t[0, 0, 0]\n",
      "3\tFalse\t18\t18\t0\t0.0\t0.0\t0.0\t[18, 0, 0, 0]\n",
      "4\tFalse\t13\t13\t0\t0.0\t0.0\t0.0\t[13, 0, 0, 0, 0]\n",
      "5\tFalse\t14\t14\t0\t0.0\t0.0\t0.0\t[14, 0, 0, 0, 0, 0]\n",
      "6\tFalse\t7\t7\t0\t0.0\t0.0\t0.0\t[7, 0, 0, 0, 0, 0, 0]\n",
      "7\tFalse\t15\t14\t1\t0.0380952380952\t0.0272108843537\t1.3416407865\t[14, 0, 0, 1, 0, 0, 0, 0]\n",
      "8\tFalse\t12\t10\t2\t0.0922619047619\t0.0642791551882\t1.62122726671\t[10, 0, 0, 1, 1, 0, 0, 0, 0]\n",
      "9\tFalse\t19\t14\t5\t0.108187134503\t0.0968255066049\t0.497044016364\t[14, 2, 0, 1, 2, 0, 0, 0, 0, 0]\n",
      "10\tFalse\t18\t16\t2\t0.0592592592593\t0.0392761958199\t1.64145266725\t[16, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0]\n",
      "11\tFalse\t28\t26\t2\t0.0298701298701\t0.0243869394391\t0.687619159566\t[26, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0]\n",
      "12\tFalse\t45\t43\t2\t0.0198653198653\t0.014717301191\t1.02214179656\t[43, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "13\tFalse\t52\t41\t11\t0.0710059171598\t0.0681676120232\t0.168305718872\t[40, 2, 3, 3, 3, 0, 0, 0, 0, 0, 0, 0, 0, 1]\n",
      "14\tFalse\t117\t101\t16\t0.04884004884\t0.0430020078657\t0.556986482607\t[99, 3, 3, 1, 7, 0, 2, 0, 0, 0, 0, 0, 0, 0, 2]\n",
      "15\tFalse\t767\t720\t47\t0.0123424598001\t0.0188456191797\t-1.48239939994\t[708, 34, 4, 2, 4, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 12]\n",
      "16\tFalse\t2663\t2474\t189\t0.0121604706471\t0.0213886948286\t-1.87807458501\t[2474, 151, 13, 6, 10, 2, 3, 3, 1, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "17\tFalse\t640\t555\t85\t0.0349954044118\t0.0392851660887\t-0.462921063556\t[552, 21, 36, 3, 13, 2, 1, 3, 6, 0, 0, 0, 0, 0, 0, 0, 0, 3]\n",
      "18\tFalse\t5257\t4755\t502\t0.0223430694959\t0.0277628338832\t-0.833414618249\t[4755, 198, 108, 72, 65, 18, 20, 7, 14, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "19\tFalse\t10\t10\t0\t0.0\t0.0\t0.0\t[10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "20\tFalse\t287\t268\t19\t0.0115349348982\t0.018660357577\t-1.44408652468\t[266, 10, 4, 4, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2]\n"
     ]
    }
   ],
   "source": [
    "hapSFSs =  SFS_from_genotypes(hapvcf, region, recordAttribute_filters=recordAttribute_filters, recordINFO_filters=recordINFO_filters, callFormat_filters=callFormat_filters, callAttribute_filters=callAttribute_filters,  ploidy=1, vcf_ploidy=1)\n",
    "for s in hapSFSs:\n",
    "     hapSFSs[s].summarise(header=False)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The strange result here is that @ first glance FREEBAYES has WAY fewer sites!!!\n",
    "*HOWEVER*, this is because it reports little haplotype chunks. \n",
    "Using the parameter '--max-complex-gap' or '--haplotype-length' you can specify the max length of these chunks\n",
    "So what happens is you might have 2-'max-complex-gap' sites all reported in a single VCF line\n",
    "This is great that FreeBayes does this - but its fucking annoying to parse!!!\n",
    "\n",
    "\n",
    "\n",
    "    freebayes -f $ref --ploidy 1 --max-complex-gap 0 --region chromosome_1:1000000-1010000 --theta 0.025 --genotype-qualities --report-monomorphic   /home2/data/sequencing/chlamydomonas/quebec/all/quebec_wt.realigned.bam >quebec_all.chromosome_1_1000000_1100000.freebayes.bpRES.vcf\n",
    "  \n",
    "  \n",
    "# Eg of a weird SNP\n",
    "\n",
    "    chromosome_1    1012696 .       TTGAATCCCCATGCGCCGGGGAGCCACGGTAGCCCGATG TTGAATCCCCATGTGCCGGGGAGCCACGGTAGCCCGATG,TTGAATCCCCATGCGCCGGGGAGCCACGGTAGCCCGATA 2383.31 .       AB=0,0;ABP=0,0;AC=1,4;AF=0.05,0.2;AN=20;AO=14,53;CIGAR=13M1X25M,38M1X;DP=280;DPB=427.923;DPRA=1,0.888889;EPP=12.937,3.05127;EPPR=3.42174;GTI=0;LEN=1,1;MEANALT=1,1.4;MQM=97.2857,98.4717;MQMR=97.9684;NS=20;NUMALT=2;ODDS=71.6637;PAIRED=1,1;PAIREDR=1;PAO=60.5,97.5;PQA=1774.17,3138.17;PQR=4459.67;PRO=141;QA=421,1541;QR=6232;RO=190;RPL=10,23;RPP=8.59409,5.01789;RPPR=3.42174;RPR=4,30;RUN=1,1;SAF=5,25;SAP=5.49198,3.37904;SAR=9,28;SRF=100;SRP=4.15318;SRR=90;TYPE=snp,snp;technology.illumina=1,1       GT:GQ:DP:RO:QR:AO:QA:GL 0:160.002:16:16:518:0,0:0,0:0,-56.7692,-56.0494 0:160.002:9:9:330:0,0:0,0:0,-34.3943,-34.231    0:160.002:11:9:277:0,0:0,0:0,-33.3755,-33.6034  0:160.002:10:9:266:0,1:0,34:0,-44.6318,-41.6855 0:160.002:15:15:490:0,0:0,0:0,-50.5683,-50.412  2:160.002:9:0:0:0,9:0,294:-42.6357,-42.4674,0   2:160.002:13:0:0:0,13:0,386:-54.3905,-54.1013,0 2:160.002:24:0:0:0,22:0,644:-65.0939,-64.6268,0 2:160.002:8:0:0:0,8:0,183:-32.912,-32.6298,0    0:160.002:21:18:616:0,0:0,0:0,-57.9422,-57.4051 1:160.002:14:0:0:14,0:421,0:-59.2619,0,-58.3784 0:160.002:19:18:588:0,0:0,0:0,-61.339,-60.7963  0:160.002:14:14:466:0,0:0,0:0,-45.5041,-45.1759 0:160.002:10:9:332:0,0:0,0:0,-31.4781,-31.7232  0:160.002:16:11:357:0,0:0,0:0,-39.6721,-39.8266 0:160.002:17:16:504:0,0:0,0:0,-48.5697,-47.9172 0:160.002:16:14:457:0,0:0,0:0,-43.4396,-43.4396 0:160.002:11:10:332:0,0:0,0:0,-32.5941,-32.0848 0:160.002:13:11:352:0,0:0,0:0,-38.467,-38.5823  0:160.002:14:11:347:0,0:0,0:0,-31.1668,-30.6165\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "- REF    TTGAATCCCCATG C GCCGGGGAGCCACGGTAGCCCGAT G\n",
    "- ALT1   TTGAATCCCCATG T GCCGGGGAGCCACGGTAGCCCGAT G\n",
    "- ALT2   TTGAATCCCCATG C GCCGGGGAGCCACGGTAGCCCGAT A\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.019540452475 -0.868736138935\n",
      "0.0223430694959 -0.833414618249\n",
      "Freebayes: 0.0169554122898\n",
      "HaplotypeCaller: 0.0201226113439\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def mean_weightedBySites(SFSs, min_alleles=1):\n",
    "    total_sites = 0.0\n",
    "    total_pi = 0.0\n",
    "    for s in SFSs:\n",
    "        if s >min_alleles:\n",
    "            try:\n",
    "                total_pi += SFSs[s].sites() * SFSs[s].theta_pi()\n",
    "                total_sites +=SFSs[s].sites()\n",
    "            except TypeError:\n",
    "                continue\n",
    "    return total_pi/total_sites\n",
    "    \n",
    "print freeSFSs[20].theta_pi(), freeSFSs[20].taj_D()\n",
    "print hapSFSs[18].theta_pi(), hapSFSs[18].taj_D()\n",
    "\n",
    "print \"Freebayes:\",  mean_weightedBySites(freeSFSs, 8)\n",
    "print \"HaplotypeCaller:\", mean_weightedBySites(hapSFSs,  8)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
