{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam   \n",
    "import collections\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from cyvcf2 import VCF\n",
    "\n",
    "df_breakpoints = pd.read_csv('../../data/references/liu-data-reference/all_break_points.txt', sep='\\t')\n",
    "df_lookup = pd.read_csv('../../data/references/liu-data-reference/liu_sample_lookup.tsv', sep='\\t')\n",
    "bam_files_path = '../../data/liu-data/bam/'\n",
    "vcf_files_path = '../../data/liu-data/vcf/'\n",
    "reference_files_path = '../../data/references/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Grab reads from all_break_points.txt, look up their corresponding bam file in liu_sample_lookup.tsv\n",
    "and write the reads into a text file\n",
    "'''\n",
    "\n",
    "with open('breakpoint_bams.txt', 'w') as f:\n",
    "    for index, row in df_breakpoints.iterrows():\n",
    "        \n",
    "        individual = row[2]\n",
    "        chromosome = row[3]\n",
    "        left_bound = int(row[4])\n",
    "        right_bound = int(row[5])\n",
    "        \n",
    "        try:\n",
    "            run = df_lookup[df_lookup['individual'] == individual].iloc[0]['Run'] + '.dup.bam'\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "        alignment_file = pysam.AlignmentFile(bam_files_path + run, 'rb')\n",
    "        for alignment in alignment_file.fetch(chromosome, left_bound, right_bound):\n",
    "            f.write(str(alignment)+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Using cyvcf, go through each of the breakpoints and check the number of SNPs per segment.\n",
    "Return the SNP locations in a list.\n",
    "'''\n",
    "\n",
    "def check_snps(f_name, chromosome, left_bound, right_bound):\n",
    "    vcf_in = VCF(f_name)\n",
    "    region = '{c}:{l}-{r}'.format(c=chromosome, l=left_bound, r=right_bound)\n",
    "    records = [rec for rec in vcf_in(region)]\n",
    "    return records\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "SNP distribution analysis per 10k bases\n",
    "'''\n",
    "\n",
    "bound = 0\n",
    "SNP_count = []\n",
    "f_name = vcf_files_path + 'parental_filtered.vcf.gz'\n",
    "\n",
    "with open('SNP_10k.txt', 'w') as f:\n",
    "    while True:\n",
    "        count = len(check_SNPs(f_name, 'chromosome_1', bound, bound + 10000))\n",
    "        SNP_count.append(count)\n",
    "        f.write('{b1}-{b2}: {c}\\n'.format(b1=bound, b2=bound+10000, c=count))\n",
    "        bound += 10000\n",
    "        \n",
    "        #didn't find a way to get the length of sequence or last SNP position, so this is my hack\n",
    "        if SNP_count[-5:] == [0, 0, 0, 0, 0]:\n",
    "            break\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "SNP distribution analysis per breakpoint\n",
    "'''\n",
    "f_name = vcf_files_path + 'parental_filtered.vcf.gz'\n",
    "\n",
    "with open('SNP_breakpoint.txt', 'w') as f:\n",
    "    for index, row in df_breakpoints.iterrows():\n",
    "        \n",
    "        individual = row[2]\n",
    "        chromosome = row[3]\n",
    "        left_bound = int(row[4])\n",
    "        right_bound = int(row[5])\n",
    "        \n",
    "        try:\n",
    "            run = df_lookup[df_lookup['individual'] == individual].iloc[0]['Run'] + '.dup.bam'\n",
    "        except:\n",
    "            continue\n",
    "        \n",
    "        #parental_filtered.vcf only has chromosome 1\n",
    "        if chromosome != \"chromosome_1\":\n",
    "            continue\n",
    "        \n",
    "        count = len(check_SNPs(f_name, chromosome, left_bound, right_bound))\n",
    "        f.write('{run} {chrom} {l} {r} {c}\\n'.format(run = run, chrom=chromosome, l=left_bound, r=right_bound, c=count))\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Create bam file using first region of all_break_points.txt. 10% of reads should be of that section and the\n",
    "rest should be from other individuals without the recombination there\n",
    "\n",
    "recomb_filepath: full filepath to the bam file that the recombination is on\n",
    "    ex: bam_file_path + 'SRR5243261.dup.bam'\n",
    "chromsome: chromosome that the recombination is on\n",
    "    ex: chromsome_1\n",
    "left_bound/right_bound: the bounds of the recombination\n",
    "    ex: 3440560, 3441577\n",
    "lines: lines from each files, total number of lines in mock file is lines * (no. of other_files + 1)\n",
    "    Note: there may not be enough reads for your given amount of lines\n",
    "other_files: default is set in function, can also be user defined\n",
    "\n",
    "'''\n",
    "\n",
    "def recomb_mock(recomb_filepath, chromosome, left_bound, right_bound, lines=200, other_files=None):\n",
    "    \n",
    "    if other_files == None:\n",
    "        other_files = ['SRR5243250.dup.bam', 'SRR5243251.dup.bam', 'SRR5243252.dup.bam', 'SRR5243253.dup.bam', \n",
    "             'SRR5243254.dup.bam', 'SRR5243255.dup.bam', 'SRR5243256.dup.bam', 'SRR5243257.dup.bam', 'SRR5243262.dup.bam']\n",
    "    \n",
    "    recomb_file = pysam.AlignmentFile(recomb_filepath)\n",
    "\n",
    "    f = pysam.AlignmentFile('recomb_mock.bam', 'wh', template=recomb_file)\n",
    "\n",
    "    counter = 0\n",
    "\n",
    "    for alignment in recomb_file.fetch(chromosome, left_bound, right_bound):\n",
    "        if counter >= lines:\n",
    "            break\n",
    "        f.write(alignment)\n",
    "        counter += 1\n",
    "        \n",
    "    print('recomb, ' +  str(counter))\n",
    "    \n",
    "    for file in other_files:\n",
    "        alignment_file = pysam.AlignmentFile(bam_files_path + file)\n",
    "\n",
    "        counter = 0\n",
    "\n",
    "        for alignment in alignment_file.fetch(chromosome, left_bound, right_bound):\n",
    "\n",
    "            # write 200 lines from each file\n",
    "            if counter >= lines:\n",
    "                break\n",
    "\n",
    "            f.write(alignment)\n",
    "            counter += 1\n",
    "        \n",
    "        print(file, counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recomb, 1449\n",
      "SRR5243250.dup.bam 1509\n",
      "SRR5243251.dup.bam 1623\n",
      "SRR5243252.dup.bam 1328\n",
      "SRR5243253.dup.bam 1323\n",
      "SRR5243254.dup.bam 1368\n",
      "SRR5243255.dup.bam 1810\n",
      "SRR5243256.dup.bam 1501\n",
      "SRR5243257.dup.bam 1449\n",
      "SRR5243262.dup.bam 1643\n"
     ]
    }
   ],
   "source": [
    "filepath = bam_files_path + 'SRR5243277.dup.bam'\n",
    "\n",
    "f = open(filepath)\n",
    "\n",
    "#recomb_mock(filepath, 'chromosome_1', 3440560, 3441577)\n",
    "#recomb_mock(filepath, 'chromosome_1', 3440400, 3441700)\n",
    "recomb_mock(filepath, 'chromosome_1', 1488300, 1497200, lines=2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Given a bam file object, return the percentage of recombination in the bam files\n",
    "\n",
    "If bam=True, create a new bam file with reads, if not, print in a human readable form\n",
    "\n",
    "If mode='no_match', create a file with only no_match sequences.\n",
    "If mode='all', create a file with all sequences\n",
    "If mode='phase_change', create a file with only sequences with phase changes\n",
    "'''\n",
    "\n",
    "def recomb_diagnosis(bam_file_obj, bam=False, mode='no_match', output_filename='recomb_diagnosis'):\n",
    "    \n",
    "    '''\n",
    "    recomb = 0\n",
    "    no_match = 0\n",
    "    total = 0\n",
    "    SNP_count = 0\n",
    "    skip_counter = 0\n",
    "    '''\n",
    "    \n",
    "    if bam:\n",
    "        f_obj = pysam.AlignmentFile(output_filename + '.sam', 'wh', template=bam_file_obj)\n",
    "        \n",
    "    else:\n",
    "        f_obj = open(output_filename + '.txt', 'w')\n",
    "\n",
    "        f_obj.write('Key: \\nSequence: Start - End \\nReference sequence \\nPhred Scale Quality \\nQuery alignment sequence \\n')\n",
    "        f_obj.write('1: CC2935, 2: CC2936, N: Does not match SNP \\n\\n')\n",
    "\n",
    "    # get reference segment\n",
    "    seq_obj = SeqIO.parse(reference_files_path + 'chlamy.5.3.w_organelles_mtMinus.fasta', 'fasta')\n",
    "\n",
    "    # grab chromosome 1\n",
    "    for seq in seq_obj:\n",
    "        chrom_1 = seq\n",
    "        break    \n",
    "    \n",
    "    for record in bam_file_obj:\n",
    "                    \n",
    "        snps = check_snps(vcf_files_path + 'parental_filtered.vcf.gz', record.reference_name, \n",
    "                          record.reference_start + 1,\n",
    "                          record.reference_start + record.query_alignment_length + 1)\n",
    "        \n",
    "        '''\n",
    "        # double list comp to see if the first of all the cigartuples are 0 or 8\n",
    "        # 0 is match, 8 is point mismatch\n",
    "        if not all([n in [0, 8] for n in [t[0] for t in record.cigartuples]]):\n",
    "            skip_counter += 1\n",
    "            continue\n",
    "        '''\n",
    "        \n",
    "        if len(snps) > 1:\n",
    "            \n",
    "            # tuple-checking\n",
    "            cigar_tuples = record.cigartuples\n",
    "            \n",
    "            # initialize segment for building\n",
    "            segment = ''\n",
    "            \n",
    "            # reference sequence\n",
    "            ref = chrom_1[record.reference_start:record.reference_start + record.query_alignment_length] \n",
    "            \n",
    "            # flag to keep track of whether or not there is an insertion in the cigar string\n",
    "            insertion_flag = False\n",
    "            \n",
    "            # 0 is match and sequences are length of 150, so it's a full match\n",
    "            if cigar_tuples == [(0, 150)]:\n",
    "                segment = record.query_sequence\n",
    "            else:\n",
    "                # index to keep track of where we are in the query_segment\n",
    "                query_segment = record.query_sequence\n",
    "                index = 0\n",
    "                \n",
    "                for cigar_tuple in cigar_tuples:\n",
    "                    # 4 = soft clipping, the record.query_sequence has the portion that is soft clipping\n",
    "                    # so we need to skip it with index\n",
    "                    if cigar_tuple[0] == 4:\n",
    "                        index += cigar_tuple[1]\n",
    "                        \n",
    "                    # 5 = hard clipping, record.query_sequence does not have the portion that is\n",
    "                    # hard clipping so we don't skip it and we don't add anything\n",
    "                    elif cigar_tuple[0] == 5:\n",
    "                        continue\n",
    "                    \n",
    "                    # if it is a match(0), then just add it onto the segment \n",
    "                    elif cigar_tuple[0] == 0:\n",
    "                        segment += query_segment[index:index+cigar_tuple[1]]\n",
    "                        index += cigar_tuple[1]\n",
    "                    \n",
    "                    # 1 is an insertion, we will add gaps to the reference\n",
    "                    elif cigar_tuple[0] == 1:\n",
    "                        segment += query_segment[index:index+cigar_tuple[1]]\n",
    "                        index += cigar_tuple[1]\n",
    "                        \n",
    "                        ref = ref[:index] + '-' * cigar_tuple[1] + ref[index:]\n",
    "                        \n",
    "                        insertion_flag = True\n",
    "                        \n",
    "                    else:\n",
    "                        print('oops forgot to consider this: ' + str(cigar_tuple))\n",
    "                        print(cigar_tuples)\n",
    "            \n",
    "            previous = 'not set'\n",
    "            \n",
    "            snp_lst = [' '] * record.query_alignment_length\n",
    "            \n",
    "            # no match flag, is true when we detect a no match for one of the SNPs below\n",
    "            no_match_flag = False\n",
    "            \n",
    "            # flag for phase detection\n",
    "            phase_change_flag = False\n",
    "            previous_strand = None\n",
    "            \n",
    "            for snp in snps:\n",
    "                # Using SNP.start and record.reference_start since they are both 0 based\n",
    "                \n",
    "                start = snp.start - record.reference_start\n",
    "                \n",
    "                # extra calculations to realign start if there is an insertion\n",
    "                if insertion_flag:\n",
    "                    \n",
    "                    current_tuple = 0\n",
    "                    current_base = 0\n",
    "                    \n",
    "                    while current_base < start and current_tuple < len(cigar_tuples):\n",
    "                        if cigar_tuples[current_tuple][0] == 1:\n",
    "                            # shift the start over by the amount of insertion to compensate for it\n",
    "                            start += cigar_tuples[current_tuple][1]\n",
    "                        \n",
    "                        current_base += cigar_tuples[current_tuple][1]\n",
    "                        current_tuple += 1\n",
    "                            \n",
    "                        \n",
    "                \n",
    "                # indexing for VCF seems to be a bit weird and will sometimes be -1\n",
    "                if start < 0:\n",
    "                    raise Exception('VCF indexing is off. Check SNP at {}'.format(snp))\n",
    "                \n",
    "                strand1 = snp.gt_bases[0][0]\n",
    "                strand2 = snp.gt_bases[1][0]\n",
    "                \n",
    "                if start >= len(segment):\n",
    "                    break\n",
    "                \n",
    "                if segment[start] == strand1:\n",
    "                    snp_lst[start] = '1'\n",
    "                    \n",
    "                    # phase change detection\n",
    "                    if previous_strand == 'strand2':\n",
    "                        phase_change_flag = 'True'\n",
    "                    previous_strand = 'strand1'\n",
    "                    \n",
    "                elif segment[start] == strand2:\n",
    "                    snp_lst[start] = '2'\n",
    "                    \n",
    "                    if previous_strand == 'strand1':\n",
    "                        phase_change_flag = 'True'\n",
    "                    previous_strand = 'strand2'\n",
    "                else:\n",
    "                    snp_lst[start] = 'N'\n",
    "                    no_match_flag = True\n",
    "                    \n",
    "            snp_str = ''.join(snp_lst)\n",
    "            \n",
    "            # qualities\n",
    "            qualities_str = ''\n",
    "            for quality in record.query_alignment_qualities:\n",
    "                qualities_str += chr(quality + 33)\n",
    "            \n",
    "            \n",
    "            # we only want no_match and we didn't find a no_match\n",
    "            if mode == 'no_match' and not no_match_flag:\n",
    "                continue\n",
    "            \n",
    "            elif mode == 'phase_change' and not phase_change_flag:\n",
    "                continue\n",
    "            \n",
    "            elif bam:\n",
    "                f_obj.write(record)\n",
    "                \n",
    "            else:\n",
    "                f_obj.write('Sequence: {start} - {end} \\n'.format(start=record.reference_start, \n",
    "                                                               end=record.reference_start+record.query_alignment_length))\n",
    "\n",
    "                f_obj.write(str(cigar_tuples) + '\\n')\n",
    "\n",
    "                f_obj.write(str(ref.seq) + '\\n' + qualities_str + '\\n' + segment + '\\n' + snp_str + '\\n \\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/site-packages/ipykernel_launcher.py:139: DeprecationWarning: In future, it will be an error for 'np.bool_' scalars to be interpreted as an index\n",
      "/usr/local/lib/python3.5/site-packages/ipykernel_launcher.py:140: DeprecationWarning: In future, it will be an error for 'np.bool_' scalars to be interpreted as an index\n"
     ]
    }
   ],
   "source": [
    "bam_file_obj = pysam.AlignmentFile('recomb_mock.bam', 'r')\n",
    "\n",
    "recomb_diagnosis(bam_file_obj, bam=False, mode='phase_change')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
